### ISSUE TITLE
Bug in expand terms

### ISSUE THREAD

**Issue by yossigil** [SpartanRefactoring/Spartanizer#129]
The TermsExpander get confused by this:
```
      if (finalCount <= 0)
        return announce("Spartanizing '" + javaProject.getElementName() + "' project \n" + //
            "Completed in " + (1 + i) + " passes. \n" + //
            "Total changes: " + (initialCount - finalCount) + "\n" + //
            "Suggestions before: " + initialCount + "\n" + //
            "Suggestions after: " + finalCount + "\n" + //
            message);
    }
```

You mission should you accept is

- [x]  Implement a test case for this example, showing tha the trimmer gets confused
- [x]  Idenitfy the source of the problem
- [x]  make more test cases demonstrating the problem
- [x] solve the problem with the class you just wrote with Niv

**Comment by yossigil** [comment]
Pending on issue #119

**Comment by yossigil** [comment]
@NivShalmon I think you are the types expert

**Comment by NivShalmon** [comment]
As I explained in #119, this is probably caused since PrudentType thinks (initialCount - finalCount) is a NUMERIC and than TermExpandor works on it. Still need to test this to make sure this is realy what happens

**Comment by NivShalmon** [comment]
@yossigil I think the problem here can be solved rather easily, if TermCollector will return an empty collector for InfixExpression whose type is STRING, like it does for leaf expressions. I belibe TermExpander/Reorgnizer aren't supposed to work on those cases anyway

**Comment by NivShalmon** [comment]
I looked into this a bit more and infixAdditionSort already does this for TermReorgenizer (using stringType), we should simply do the same for TermExpander, since we don't want to remove parentheses here. The downside with this idea is that cases such as `"abc" + ((x))` will not be simplified as well.

**Comment by NivShalmon** [comment]
I think this is currently a bigger problem than we understood. Look at this test:
```
 @Test public void issue72pg() {
    trimming("0+(x+y)").to("0+x+y").to("x+y").to(null);
  }
```
if x is an int whose value is 3, and y is an empty string, then the original expression evaluates to "03" while the second and third expression evaluate to "3".
I think that as long as we don't know that the expression isn't a String (stringType.isnot() doesn't return true) we can't expand/reorder the expression.
What do you think @yossigil?

**Comment by yossigil** [comment]
you hit the nail right on the head. that's what we need to do!

**Comment by yossigil** [comment]
Good job. Thanks. If done, please close

**Comment by yossigil** [comment]
@NivShalmon :+1:  Do you suppose you can resolve this issue now?

**Comment by NivShalmon** [comment]
This issue should be done. I also solved another bug in removal of 0 from infix plus, where sometimes a 0 was removed when we couldn't know if it was addition or String concating, such as `0+x` turning into `x`.
This cuases another problem however, for exmaple in this case:
```
0+(0+x+y+(4+0))
```
turns to
```
0+(0+x+y+(4))
```
and than doesn't change. The parenthesis can be safely removed, but isn't because TermExpander used to do this but can't be called here. I think this will require it's own wring.
